### DISCUSSION TITLE
layouts stored twice?

### DISCUSSION THREAD

**Discussion by TornaxO7** [leftwm/leftwm#913] [Q&A]
Why do we have the attribute [`layouts` (State)](https://github.com/leftwm/leftwm/blob/main/leftwm-core/src/state.rs#L25) and [`layouts` (LayoutManager)](https://github.com/leftwm/leftwm/blob/main/leftwm-core/src/models/layout_manager.rs#L21) stored twice?

**Comment by VuiMuich** [comment]
Layout_manager has no direct access to the state, so copying the layouts from the config was probably just more convenient to implement.

In general I guess this could be refactored, if it doesn't get obsolete with the upcoming `leftwm-layouts` library anyways.
Tagging @hertg for opinion and if the lib does not rid this, maybe he can do the refactor, while touching this part of code already.

**Reply by TornaxO7** [reply]
> In general I guess this could be refactored, if it doesn't get obsolete with the upcoming leftwm-layouts library anyways.

huh, there's such a library planned?

**Reply by hertg** [reply]
> huh, there's such a library planned?

[yes](https://github.com/leftwm/leftwm-layouts), it's pretty much finished apart from some minor stuff. I just need to find the time to actually integrate it into leftwm.

**Reply by TornaxO7** [reply]
What's the goal of this library?

**Reply by hertg** [reply]
**Refactoring**  <sub>but mainly decoupling layouts from windows and x11, reducing code duplication, possibility to calculate window positions without actually re-positioning them, allow changing layout definitions via configuration in the future, fixing #96, #516, etc</sub>

**Reply by TornaxO7** [reply]
@hertg so this binary handles the layouts then, am I right?

**Reply by hertg** [reply]
It's not a binary, it's a library. But essentially yes.

**Reply by TornaxO7** [reply]
Hm... ok, so this layout-thing is already taken care of?
